crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -652,6 +503,372 @@ where
+        tx_data: TxData,
+        mut tx_logs: TxLogs,
+    ) {
+        // Check the commitment of the fee unshielding regardless of the
+        // result, it could be committed even in case of errors
+        if tx_data
+            .wrapper_args
+            .as_ref()
+            .map(|args| args.is_committed_fee_unshield)
+            .unwrap_or_default()
+        {
+            tx_logs
+                .tx_event
+                .extend(ValidMaspTx((tx_data.tx_index, None)));
+        }
+
+        match dispatch_result {
+            Ok(()) => self.handle_inner_tx_results(
+                response,
+                tx_result,
+                tx_data,
+                &mut tx_logs,
+            ),
+            Err(Error::TxApply(protocol::Error::WrapperRunnerError(msg))) => {
+                tracing::info!(
+                    "Wrapper transaction {} failed with: {}",
+                    tx_logs.tx_event["hash"],
+                    msg,
+                );
+                tx_logs
+                    .tx_event
+                    .extend(WithGasUsed(tx_data.consumed_gas))
+                    .extend(Info(msg.to_string()))
+                    .extend(Code(ResultCode::InvalidTx));
+                // Make sure to clean the write logs for the next transaction
+                self.state.write_log_mut().drop_tx();
+            }
+            Err(msg) => {
+                // This branch represents an error that affects the entire
+                // batch
+                tracing::info!(
+                    "Transaction {} failed with: {}",
+                    tx_logs.tx_event["hash"],
+                    msg
+                );
+
+                tx_logs
+                    .tx_event
+                    .extend(WithGasUsed(tx_data.consumed_gas))
+                    .extend(Info(msg.to_string()))
+                    .extend(Code(ResultCode::WasmRuntimeError));
+
+                self.handle_batch_error(
+                    response,
+                    msg,
+                    tx_result,
+                    tx_data,
+                    &mut tx_logs,
+                );
+            }
+        }
+    }
+
+    // Evaluate the results of all the transactions of the batch. Commit or drop
+    // the storage changes, update stats and event, manage replay protection.
+    fn handle_inner_tx_results(
+        &mut self,
+        response: &mut shim::response::FinalizeBlock,
+        tx_result: namada::tx::data::TxResult<protocol::Error>,
+        tx_data: TxData,
+        tx_logs: &mut TxLogs,
+    ) {
+        // Track the need to commit the batch hash for replay protection. Hash
+        // must be written if at least one of the txs in the batch requires so
+        let mut commit_batch_hash = false;
+        // Track if any of the inner txs failed or was rejected
+        let mut is_any_tx_invalid = false;
+        let mut temp_log = TempTxLogs::default();
+
+        temp_log.check_inner_results(
```

Comment:
This is a nit, but I much prefer returning results rather than passing mutable references as arguments.

Reply:
Done in 525d044c624cc4c189a845ed96eb5b7c584715b3